Add Kaniko Lab

kn71026 opened the pull request:
This commit add Kaniko lab. In the Kaniko lab, we walk through the  
introduction of Kaniko and how to use Kaniko to build image on Kubernetes and Gitlab.

ching-kuo commented on self-paced-labs/image_build/kaniko_lab/README.md:
```diff
@@ -0,0 +1,203 @@
+    * 藍色的是每個 stage
+    * 綠色的是執行 Dockerfile 裡面的 instruction
+
+![](https://imgur.com/4cxuRFG.jpg)
```
move image from imgur to a separate folder (or directly link to reference)

ching-kuo commented on self-paced-labs/image_build/kaniko_lab/scripts/volume.yaml:
```diff
@@ -0,0 +1,14 @@
+    - ReadWriteOnce
+  storageClassName: local-storage
+  hostPath:
+    path: /home/docker/kaniko # replace with local directory, such as "/home/<user-name>/kaniko"
```
我不太確定如果 directory 不存在的時候會不會 work
如果是個 multi worker node 的 Kubernetes cluster，是不是需要確保每一個 worker node 都存在這個 hostPath?

hwchiu replied:
不存在不會 work

ching-kuo commented on self-paced-labs/image_build/kaniko_lab/scripts/.gitlab-ci.yml:
```diff
@@ -0,0 +1,18 @@
+stages:
+  - build
+build:
+  stage: build
+  image:
+    name: gcr.io/kaniko-project/executor:debug
+    entrypoint: [""]
+  script:
+    - mkdir -p /kaniko/.docker
+    - echo "{\"auths\":{\"${CI_REGISTRY}\":{\"auth\":\"$(printf "%s:%s" "${CI_REGISTRY_USER}" "${CI_REGISTRY_PASSWORD}" | base64 | tr -d '\n')\"}}}" > /kaniko/.docker/config.json
+    - >-
+      /kaniko/executor
+      --context "${CI_PROJECT_DIR}"
+      --dockerfile "${CI_PROJECT_DIR}/Dockerfile"
+      --destination "${CI_REGISTRY_IMAGE}:${CI_COMMIT_SHORT_SHA}"
+  
+  tags:
+      - gitlab-runner-k8s # 前面架設的 gitlab runner 的 tag
```
Why make this file hidden?

hwchiu replied:
the CI configuration for gitlab, it should be name as `.gitlab-ci.yml`

ching-kuo commented on self-paced-labs/image_build/kaniko_lab/README.md:
```diff
@@ -0,0 +1,203 @@
+> 理論上可以
+## 運作原理
+- 以下操作皆在 user-space 使用 executor image 執行，避免使用privileged mode，因此也不會需要 docker daemon。
+- Kaniko 會從 Base image 提取 file system
+- 當逐行執行 Dockerfile 時，在 user-space 對 filesystem 做快照，運用 cache 去和 base image 比對
+- 將有更改的 files append 到 base image，並更新 image 的 metadata
+- 當最後執行完整個 Dockerfile 時， Kaniko executor 將 build 完成的 image push 到指定的 registry 中。
```
indent this part

ching-kuo reviewed (changes requested):
Some comment and questions

hwchiu commented on self-paced-labs/image_build/kaniko_lab/README.md:
```diff
@@ -0,0 +1,203 @@
+- 在 Kaniko container 裡面 build time 要用 root 執行
+- 並非所有 Dockerfile 都適用->目前還沒看到不能用的範例
+- 較傳統用 docker build的方式複雜
+- 沒有支援local的layer caching
```
英文前後空白，跟前後用法一致

steveyiyo commented on self-paced-labs/image_build/kaniko_lab/README.md:
```diff
@@ -0,0 +1,203 @@
+
+
+### 不需要 privileged mode 的原因
+- Kaniko 在所有操作中沒有引入任何Docker daemon 或者 CLI 操作。
```
中文跟英文中間需要加一個空格

kn71026 committed:
feat: introduction Kaniko

This commit add Kaniko lab. In the Kaniko lab, we walk through the
introduction of Kaniko and how to use Kaniko to build image on Kubernetes.

steveyiyo commented on self-paced-labs/image_build/kaniko_lab/README.md:
```diff
@@ -0,0 +1,200 @@
+    * 藍色的是每個 stage
+    * 綠色的是執行 Dockerfile 裡面的 instruction
+
+![kaniko_workflow](image/kaniko_workflow.jpg)
```
Please use the relative URL to frame the image.
`![kaniko_workflow](../image/kaniko_workflow.jpg)`

<img width="334" alt="CleanShot 2022-06-20 at 23 40 51@2x" src="https://user-images.githubusercontent.com/43361940/174637422-462a140e-d997-4c5c-922f-f48debe5712a.png">

kn71026 replied:
ok, fixed.

ching-kuo reviewed (approved):
LGTM

pichuang commented on self-paced-labs/image_build/kaniko_lab/Gitlab-kaniko-lab.md:
````diff
@@ -0,0 +1,149 @@
+```shell=
+$ helm repo add gitlab https://charts.gitlab.io
+```
+若是使用的是 Helm 2，則需要先初始化 Helm
````
Suggest to remove Helm 2. No longer use it.

pichuang commented on self-paced-labs/image_build/kaniko_lab/Gitlab-kaniko-lab.md:
````diff
@@ -0,0 +1,149 @@
+- 完成後可執行
+
+```shell=
+# For Helm 2
````
Suggest to remove Helm 2. No longer use it.

pichuang commented on self-paced-labs/image_build/kaniko_lab/Gitlab-kaniko-lab.md:
````diff
@@ -0,0 +1,149 @@
+
+- 登入 Docker hub
+```
+docker login
````
Better to use `docker login index.docker.io`

pichuang commented on self-paced-labs/image_build/kaniko_lab/README.md:
```diff
@@ -0,0 +1,200 @@
+  - 在執行器鏡像（executor image）的用戶空間中對文件系統做快照（snapshot）
+
+### 優點
+- 不需 depend on docker daemon
```
`不需相依於 Docker daemon`

pichuang commented on self-paced-labs/image_build/kaniko_lab/README.md:
```diff
@@ -0,0 +1,200 @@
+### 優點
+- 不需 depend on docker daemon
+- 都在 userspace 執行
+- run as unprivileged container
```
`採用非特權權限的容器`

pichuang commented on self-paced-labs/image_build/kaniko_lab/README.md:
```diff
@@ -0,0 +1,200 @@
+
+### 優點
+- 不需 depend on docker daemon
+- 都在 userspace 執行
```
Linux Userspace? or another userspace?

pichuang commented on self-paced-labs/image_build/kaniko_lab/README.md:
```diff
@@ -0,0 +1,200 @@
+
+## 使用情境
+- 官方文件中，推薦了四種運行 Kaniko 的方法：
+    - In Kubernetes cluster
```
`In` 可以拿掉，因為是贅字

pichuang commented on self-paced-labs/image_build/kaniko_lab/README.md:
```diff
@@ -0,0 +1,200 @@
+    * S3 Bucket
+    * Azure Blob Storage
+    * Local Directory
+    * Local Tar
```
Local Tarball

pichuang commented on self-paced-labs/image_build/kaniko_lab/scripts/pod.yaml:
```diff
@@ -0,0 +1,27 @@
+    image: gcr.io/kaniko-project/executor:latest
+    args: ["--dockerfile=/workspace/dockerfile",
+            "--context=dir://workspace",
+            "--destination=<user-name>/<repo-name>"] # replace with your dockerhub account
```
Suggest to use Upper Case for highlight variable, like <USER-NAME>/<REPO-NAME> instead of <user-name>/<repo-name>

pichuang commented on self-paced-labs/image_build/kaniko_lab/README.md:
````diff
@@ -0,0 +1,200 @@
+
+#### pull image 下來測試是否能運行
+```shell=
+docker run -it <user-name>/<repo-name>
````
`docker run -it kn71026/kaniko:latest`

pichuang commented on self-paced-labs/image_build/kaniko_lab/README.md:
```diff
@@ -0,0 +1,200 @@
+這邊需要準備三個檔案
+- pod.yaml：建立執行 kaniko 的 pod
+    - `--context`： 可指定 Git Repository 作為 build context，若未特別指定則預設使用 local directory，用法如下：
+      - `--context=git://[repository url][#reference][#commit-id]`
```
多一個空格

pichuang commented on self-paced-labs/image_build/kaniko_lab/README.md:
```diff
@@ -0,0 +1,200 @@
+    - In Google Cloud Build
+    - In Docker
+- 適合在無法獲取 docker daemon 的環境下 build image：如 Kubernetes cluster、Google Kubernetes Engine
+- 適合用在 Gitlab ci 過程中去 build image
```
`CI`

pichuang commented on self-paced-labs/image_build/kaniko_lab/scripts/volume-claim.yaml:
```diff
@@ -0,0 +1,11 @@
+    - ReadWriteOnce
+  resources:
+    requests:
+      storage: 8Gi
```
10Gi

pichuang reviewed (commented):
Comment by pichuang

kn71026 committed:
feat: add Kaniko GitLab lab

This commit introduce how to use Kaniko in GitLab, and how to run GitLab runner with K8s executor.
